base/shell/explorer/taskswnd.cpp

```diff
@@ -678,7 +982,147 @@ class CTaskSwitchWnd :
+
+        tbBtn.iBitmap = TaskGroup->IconIndex;
+        tbBtn.fsState = TBSTATE_ENABLED | TBSTATE_ELLIPSES;
+        tbBtn.fsStyle = BTNS_CHECK | BTNS_NOPREFIX | BTNS_SHOWTEXT | BTNS_DROPDOWN | BTNS_WHOLEDROPDOWN;
+        tbBtn.dwData = -1;
+        tbBtn.idCommand = iIndex;
+
+        m_TaskBar.BeginUpdate();
+
+        if (!m_TaskBar.InsertButton(iIndex, &tbBtn))
+        {
+            m_TaskBar.EndUpdate();
+
+            return FALSE;
+        }
+
+        UpdateIndexesAfter(iIndex, TRUE);
+
+        TaskGroup->Index = iIndex;
+        m_ButtonCount++;
+
+        /* Update button sizes and fix the button wrapping */
+        UpdateButtonsSize(TRUE);
+
+        TaskGroup->IsCollapsed = TRUE;
+
+        m_TaskBar.EndUpdate();
+
+        return TRUE;
+    }
+
+    VOID CollapseOrExpand(BOOL bAdding)
+    {
+        INT iMaxButtons = GetMaxButtons();
+
+        BOOL bIsLower = m_ButtonCount < iMaxButtons;
+        BOOL bIsGreater = bAdding ? (m_ButtonCount >= iMaxButtons) : (m_ButtonCount > iMaxButtons);
+
+        if (m_IsGroupingEnabled && (bIsLower || bIsGreater))
+        {
+            while (bIsGreater ? (m_ButtonCount >= iMaxButtons) :
+                  (m_ButtonCount < iMaxButtons))
+            {
+                // We expanded the taskbar size, we need to collapse if grouping is enabled
+
+                PTASK_GROUP CurrentGroup = m_TaskGroups;
+                PTASK_GROUP MaxOrMinGroup = NULL;
+                while (CurrentGroup != NULL)
+                {
+                    if (bIsGreater != CurrentGroup->IsCollapsed &&
+                        bIsGreater == (CurrentGroup->Index < 0) &&
+                        CurrentGroup->dwTaskCount > 1)
+                    {
+                        if (MaxOrMinGroup == NULL ||
+                           (bIsGreater && MaxOrMinGroup->dwTaskCount < CurrentGroup->dwTaskCount) ||
+                           (bIsLower && MaxOrMinGroup->dwTaskCount > CurrentGroup->dwTaskCount))
+                           MaxOrMinGroup = CurrentGroup;
+                    }
+
+                    CurrentGroup = CurrentGroup->Next;
+                }
+
+                if (MaxOrMinGroup == NULL)
+                    return; // We can't collapse more groups
+
+                if (bIsGreater)
+                {
+                    CollapseTaskGroup(MaxOrMinGroup);
+                }
+                else if (MaxOrMinGroup->dwTaskCount <= (DWORD)(iMaxButtons - m_ButtonCount + 1))
+                {
+                    ExpandTaskGroup(MaxOrMinGroup);
+                }
+                else
+                {
+                    return;
+                }
+            }
+        }
+        else if (!m_IsGroupingEnabled)
+        {
+            // Expand all the groups
+
+            PTASK_GROUP CurrentGroup = m_TaskGroups;
+            while (CurrentGroup != NULL)
+            {
+                if (CurrentGroup->IsCollapsed)
+                {
+                    /* Collapse the group */
```

Comment:
"Expand the group", you mean? (technically the comment looks redundant since you say above: "Expand all the groups", which explains what the whole while-if-... block does.

Reply:
Yeah, it's a typo